shell/platform/linux/public/flutter_linux/fl_method_channel.h

```diff
@@ -0,0 +1,253 @@
+ * end of the channel.
+ *
+ * Returns: a new #FlMethodChannel.
+ */
+FlMethodChannel* fl_method_channel_new(FlBinaryMessenger* messenger,
+                                       const gchar* name,
+                                       FlMethodCodec* codec);
+
+/**
+ * fl_method_channel_set_method_call_handler:
+ * @channel: an #FlMethodChannel.
+ * @handler: function to call when a method call is received on this channel.
+ * @user_data: (closure): user data to pass to @handler.
+ *
+ * Sets the function called when a method is called. Call
+ * fl_method_channel_respond_success(), fl_method_channel_respond_error() or
+ * fl_method_channel_respond_not_implemented() when the method completes.
+ * Ownership of #FlMethodChannelResponseHandle is transferred to the caller, and
+ * the call must be responded to to avoid memory leaks.
+ */
+void fl_method_channel_set_method_call_handler(
+    FlMethodChannel* channel,
+    FlMethodChannelMethodCallHandler handler,
+    gpointer user_data);
+
+/**
+ * fl_method_channel_invoke_method:
+ * @channel: an #FlMethodChannel.
+ * @method: the method to call.
+ * @args: (allow-none): arguments to the method, must match what the
+ * #FlMethodCodec supports.
+ * @cancellable: (allow-none): a #GCancellable or %NULL.
+ * @callback: (scope async): (allow-none): a #GAsyncReadyCallback to call when
+ * the request is satisfied or %NULL to ignore the response.
+ * @user_data: (closure): user data to pass to @callback.
+ *
+ * Calls a method on this channel.
+ */
+void fl_method_channel_invoke_method(FlMethodChannel* channel,
+                                     const gchar* method,
+                                     FlValue* args,
+                                     GCancellable* cancellable,
+                                     GAsyncReadyCallback callback,
+                                     gpointer user_data);
+
+/**
+ * fl_method_channel_invoke_method_finish:
+ * @channel: an #FlMethodChannel.
+ * @result:  #GAsyncResult.
+ * @error: (allow-none): #GError location to store the error occurring, or %NULL
+ * to ignore.
+ *
+ * Completes request started with fl_method_channel_invoke_method().
+ *
+ * Returns: (transfer full): an #FlMethodResponse or %NULL on error.
+ */
+FlMethodResponse* fl_method_channel_invoke_method_finish(
+    FlMethodChannel* channel,
+    GAsyncResult* result,
+    GError** error);
+
+/**
+ * fl_method_channel_respond_success:
+ * @channel: an #FlMethodChannel.
+ * @response_handle: handle that was provided in a
+ * #FlMethodChannelMethodCallHandler.
+ * @result: (allow-none): value to respond with, must match what the
+ * #FlMethodCodec supports.
+ * @error: (allow-none): #GError location to store the error occurring, or %NULL
+ * to ignore.
+ *
+ * Responds to a method call.
+ *
+ * Returns: %TRUE on success.
+ */
+gboolean fl_method_channel_respond_success(
+    FlMethodChannel* channel,
+    FlMethodChannelResponseHandle* response_handle,
```

Comment:
Something that occurred to me while writing the above: I think it's worth revisiting the question of having a reply object to interact with, instead of having to call a method on the channel. For async handling, of messages, client code is always going to have to persist these two objects together (or pass them around together), which is a code smell. I suspect that if we don't have an object, a lot of people will need to make one themselves for anything async.

Rather than having response sending on the channel, we could have a simple object that internally knows its channel and handle, and has a 'send response' method that you call. Then only one object needs to be persisted or passed around when delegating the handling of responding to another function (especially if async).